File: src/core/parser/ast.ts
/**
 * AST Type Definitions for hone DSL
 *
 * Uses discriminated union types for TypeScript-friendly exhaustive checking.
 * All nodes include line numbers for error diagnostics.
 */

// String literal types
export type StringLiteral = {
  value: string;
  raw: string; // Original string including quotes
  quoteType: "single" | "double";
};

// Regex literal type
export type RegexLiteral = {
  pattern: string;
  flags: string;
  raw: string; // Original regex including slashes
};

// Duration units
export type DurationUnit = "ms" | "s";

// Duration value
export type Duration = {
  value: number;
  unit: DurationUnit;
  raw: string;
};

// Output selectors
export type OutputSelector = "stdout" | "stdout.raw" | "stderr";

// Comparison operators
export type ComparisonOperator = "==" | "!=" | "<" | "<=" | ">" | ">=";

// String comparison operators (subset)
export type StringComparisonOperator = "==" | "!=";

// Assertion predicates for output
export type OutputPredicate =
  | { type: "contains"; value: StringLiteral }
  | { type: "matches"; value: RegexLiteral }
  | { type: "equals"; operator: StringComparisonOperator; value: StringLiteral };

// Assertion predicates for exit code
export type ExitCodePredicate = {
  operator: StringComparisonOperator;
  value: number;
};

// Assertion predicates for duration
export type DurationPredicate = {
  operator: ComparisonOperator;
  value: Duration;
};

// Assertion predicates for files
export type FilePredicate =
  | { type: "exists" }
  | { type: "contains"; value: StringLiteral }
  | { type: "matches"; value: RegexLiteral }
  | { type: "equals"; operator: StringComparisonOperator; value: StringLiteral };

// Assertion types
export type AssertionExpression =
  | {
      type: "output";
      target?: string; // Named RUN reference
      selector: OutputSelector;
      predicate: OutputPredicate;
    }
  | {
      type: "exit_code";
      target?: string; // Named RUN reference
      predicate: ExitCodePredicate;
    }
  | {
      type: "duration";
      target?: string; // Named RUN reference
      predicate: DurationPredicate;
    }
  | {
      type: "file";
      path: StringLiteral;
      predicate: FilePredicate;
    };

// Pragma types
export type PragmaType = "shell" | "env" | "timeout" | "unknown";

export interface PragmaNode {
  type: "pragma";
  pragmaType: PragmaType;
  key?: string; // For env pragma
  value: string;
  line: number;
  raw: string;
}

export interface CommentNode {
  type: "comment";
  text: string;
  line: number;
}

export interface TestNode {
  type: "test";
  name: string;
  line: number;
}

export interface RunNode {
  type: "run";
  name?: string;
  command: string;
  line: number;
}

export interface AssertNode {
  type: "assert";
  expression: AssertionExpression;
  line: number;
  raw: string;
}

export interface EnvNode {
  type: "env";
  key: string;
  value: string;
  line: number;
}

// All possible AST nodes
export type ASTNode =
  | PragmaNode
  | CommentNode
  | TestNode
  | RunNode
  | AssertNode
  | EnvNode;

// Parsed file representation
export interface ParsedFile {
  filename: string;
  pragmas: PragmaNode[];
  nodes: ASTNode[];
  warnings: ParseWarning[];
}

// Parse warning (for unknown pragmas, etc.)
export interface ParseWarning {
  message: string;
  line: number;
  filename: string;
}

// Parse error
export interface ParseError {
  message: string;
  line: number;
  filename: string;
}

// Parse result
export type ParseResult =
  | { success: true; file: ParsedFile }
  | { success: false; errors: ParseError[]; warnings: ParseWarning[] };

File: src/parser/ast.rs
#[derive(Debug, Clone, PartialEq, Copy)]
pub struct Span {
    pub start: usize,
    pub end: usize,
    pub start_line: usize,
    pub start_col: usize,
    pub end_line: usize,
    pub end_col: usize,
}

impl Span {
    pub fn new(
        start: usize,
        end: usize,
        start_line: usize,
        start_col: usize,
        end_line: usize,
        end_col: usize,
    ) -> Self {
        Self {
            start,
            end,
            start_line,
            start_col,
            end_line,
            end_col,
        }
    }

    pub fn single_line(line: usize, start_col: usize, end_col: usize) -> Self {
        Self {
            start: 0,
            end: 0,
            start_line: line,
            start_col,
            end_line: line,
            end_col,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum QuoteType {
    Single,
    Double,
}

#[derive(Debug, Clone, PartialEq)]
pub struct StringLiteral {
    pub value: String,
    pub raw: String,
    pub quote_type: QuoteType,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RegexLiteral {
    pub pattern: String,
    pub flags: String,
    pub raw: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum DurationUnit {
    Milliseconds,
    Seconds,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Duration {
    pub value: f64,
    pub unit: DurationUnit,
    pub raw: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum OutputSelector {
    Stdout,
    StdoutRaw,
    Stderr,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ComparisonOperator {
    Equal,
    NotEqual,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
}

#[derive(Debug, Clone, PartialEq)]
pub enum StringComparisonOperator {
    Equal,
    NotEqual,
}

#[derive(Debug, Clone, PartialEq)]
pub enum OutputPredicate {
    Contains {
        value: StringLiteral,
    },
    Matches {
        value: RegexLiteral,
    },
    Equals {
        operator: StringComparisonOperator,
        value: StringLiteral,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExitCodePredicate {
    pub operator: StringComparisonOperator,
    pub value: i32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DurationPredicate {
    pub operator: ComparisonOperator,
    pub value: Duration,
}

#[derive(Debug, Clone, PartialEq)]
pub enum FilePredicate {
    Exists,
    Contains {
        value: StringLiteral,
    },
    Matches {
        value: RegexLiteral,
    },
    Equals {
        operator: StringComparisonOperator,
        value: StringLiteral,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum AssertionExpression {
    Output {
        target: Option<String>,
        selector: OutputSelector,
        predicate: OutputPredicate,
    },
    ExitCode {
        target: Option<String>,
        predicate: ExitCodePredicate,
    },
    Duration {
        target: Option<String>,
        predicate: DurationPredicate,
    },
    File {
        path: StringLiteral,
        predicate: FilePredicate,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum PragmaType {
    Shell,
    Env,
    Timeout,
    Unknown,
}

#[derive(Debug, Clone, PartialEq)]
pub struct PragmaNode {
    pub pragma_type: PragmaType,
    pub key: Option<String>, // For env pragma
    pub value: String,
    pub line: usize,
    pub raw: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CommentNode {
    pub text: String,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TestNode {
    pub name: String,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RunNode {
    pub name: Option<String>,
    pub command: String,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct AssertNode {
    pub expression: AssertionExpression,
    pub line: usize,
    pub raw: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnvNode {
    pub key: String,
    pub value: String,
    pub line: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ErrorNode {
    pub message: String,
    pub span: Span,
    pub raw: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ASTNode {
    Pragma(PragmaNode),
    Comment(CommentNode),
    Test(TestNode),
    Run(RunNode),
    Assert(AssertNode),
    Env(EnvNode),
    Error(ErrorNode),
}

impl ASTNode {
    pub fn line(&self) -> usize {
        match self {
            ASTNode::Pragma(node) => node.line,
            ASTNode::Comment(node) => node.line,
            ASTNode::Test(node) => node.line,
            ASTNode::Run(node) => node.line,
            ASTNode::Assert(node) => node.line,
            ASTNode::Env(node) => node.line,
            ASTNode::Error(node) => node.span.start_line,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ParsedFile {
    pub filename: String,
    pub pragmas: Vec<PragmaNode>,
    pub nodes: Vec<ASTNode>,
    pub warnings: Vec<ParseWarning>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ParseWarning {
    pub message: String,
    pub line: usize,
    pub filename: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ParseErrorDetail {
    pub message: String,
    pub line: usize,
    pub filename: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParseResult {
    Success {
        file: ParsedFile,
    },
    Failure {
        errors: Vec<ParseErrorDetail>,
        warnings: Vec<ParseWarning>,
    },
}

pub type HoneFile = ParsedFile;
